pkg/pf/internal/plugin/provider_server.go

```diff
@@ -76,63 +76,46 @@ func (p *providerServer) checkNYI(method string, err error) error {
 	return err
 }
 
-func (p *providerServer) marshalDiff(diff pl.DiffResult) (*pulumirpc.DiffResponse, error) {
-	changes := pulumirpc.DiffResponse_DIFF_UNKNOWN
+func (p *providerServer) marshalDiff(diff plugin.DiffResult) (*pulumirpc.DiffResponse, error) {
+	var changes pulumirpc.DiffResponse_DiffChanges
 	switch diff.Changes {
 	case pl.DiffNone:
 		changes = pulumirpc.DiffResponse_DIFF_NONE
 	case pl.DiffSome:
 		changes = pulumirpc.DiffResponse_DIFF_SOME
+	case pl.DiffUnknown:
+		changes = pulumirpc.DiffResponse_DIFF_UNKNOWN
 	}
 
 	// Infer the result from the detailed diff.
 	var diffs, replaces []string
 	var detailedDiff map[string]*pulumirpc.PropertyDiff
-	if len(diff.DetailedDiff) == 0 {
-		diffs = make([]string, len(diff.ChangedKeys))
-		for i, k := range diff.ChangedKeys {
-			diffs[i] = string(k)
-		}
-		replaces = make([]string, len(diff.ReplaceKeys))
-		for i, k := range diff.ReplaceKeys {
-			replaces[i] = string(k)
-		}
-	} else {
-		changes = pulumirpc.DiffResponse_DIFF_SOME
-
+	if len(diff.DetailedDiff) != 0 {
 		detailedDiff = make(map[string]*pulumirpc.PropertyDiff)
 		for path, diff := range diff.DetailedDiff {
-			diffs = append(diffs, path)
-
-			var kind pulumirpc.PropertyDiff_Kind
-			switch diff.Kind {
-			case pl.DiffAdd:
-				kind = pulumirpc.PropertyDiff_ADD
-			case pl.DiffAddReplace:
-				kind, replaces = pulumirpc.PropertyDiff_ADD_REPLACE, append(replaces, path)
-			case pl.DiffDelete:
-				kind = pulumirpc.PropertyDiff_DELETE
-			case pl.DiffDeleteReplace:
-				kind, replaces = pulumirpc.PropertyDiff_DELETE, append(replaces, path)
-			case pl.DiffUpdate:
-				kind = pulumirpc.PropertyDiff_UPDATE
-			case pl.DiffUpdateReplace:
-				kind, replaces = pulumirpc.PropertyDiff_UPDATE_REPLACE, append(replaces, path)
-			}
-
 			detailedDiff[path] = &pulumirpc.PropertyDiff{
-				Kind:      kind,
+				Kind:      pulumirpc.PropertyDiff_Kind(diff.Kind), //nolint:gosec
```

Comment:
This cast relies on accidental enum numbering? How are we sure there's no off-by-1 errors?